src/Data/Text/Utf8/BoyerMoore/Automaton.hs

```diff
@@ -0,0 +1,335 @@
+  deriving anyclass (NFData)
+
+-- | Lookup an entry in the suffix table.
+suffixLookup :: SuffixTable -> CodeUnitIndex -> CodeUnitIndex
+{-# INLINE suffixLookup #-}
+suffixLookup (SuffixTable table) = CodeUnitIndex . indexTable table . codeUnitIndex
+
+buildSuffixTable :: Text -> SuffixTable
+buildSuffixTable pattern = runST $ do
+  let patLen = Utf8.lengthUtf8 pattern
+
+  table <- UMVector.new $ codeUnitIndex patLen
+
+  let
+    -- Case 1: For each position of the pattern we record the shift that would align the pattern so
+    -- that it starts at the longest suffix that is at the same time a prefix, if a mismatch would
+    -- happen at that position.
+    --
+    -- Suppose the length of the pattern is n, a mismatch occurs at position i in the pattern and j
+    -- in the haystack, then we know that pattern[i+1..n] == haystack[j+1..j+n-i]. That is, we know
+    -- that the part of the haystack that we already matched is a suffix of the pattern.
+    -- If the pattern happens to have a prefix that is equal to or a shorter suffix of that matched
+    -- suffix, we can shift the pattern to the right so that the pattern starts at the longest
+    -- suffix that we have seen that conincides with a prefix of the pattern.
+    --
+    -- Consider the pattern `ababa`. Then we get
+    --
+    -- p:                0  1  2  3  4
+    -- Pattern:          a  b  a  b  a
+    -- lastPrefixIndex:  2  2  4  4  5
+    -- table:            6  5  6  5  5
+    init1 lastPrefixIndex p
+      | p >= 0 = do
+        let
+          prefixIndex
+            | isPrefix pattern (p + 1) = p + 1
+            | otherwise = lastPrefixIndex
+        UMVector.write table (codeUnitIndex p) (codeUnitIndex $ prefixIndex + patLen - 1 - p)
+        init1 prefixIndex (p - 1)
+      | otherwise = pure ()
+
+    -- Case 2: We also have to account for the fact that the matching suffix of the pattern might
+    -- occur again somewhere within the pattern. In that case, we may not shift as far as if it was
+    -- a prefix. That is why the `init2` loop is run after `init1`, potentially overwriting some
+    -- entries with smaller shifts.
+    init2 p
+      | p < patLen - 1 = do
+        let
+          suffixLen = suffixLength pattern p
+        when (Utf8.unsafeIndexCodeUnit pattern (p - suffixLen) /= Utf8.unsafeIndexCodeUnit pattern (patLen - 1 - suffixLen)) $
+          UMVector.write table (codeUnitIndex $ patLen - 1 - suffixLen) (codeUnitIndex $ patLen - 1 - p + suffixLen)
+        init2 (p + 1)
+      | otherwise = pure ()
+
+  init1 (patLen - 1) (patLen - 1)
+  init2 0
+
+  SuffixTable <$> UVector.unsafeFreeze table
+
+
+-- | The bad char table tells us how far we may skip ahead when encountering a certain character
+-- in the input string. For example, if there's a character that is not contained in the pattern at
+-- all, we can skip ahead until after that character.
+data BadCharTable = BadCharTable
+  { badCharTableEntries :: {-# UNPACK #-} !(UVector.Vector Int)
```

Comment:
Might be another interesting candidate for `TypedByteArray`.

Reply:
I'll draft up a separate PR that updates the `SuffixTable` as well and applies the changes to the UTF-16 version too.